src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScannerTest.java

```diff
@@ -1027,4 +1028,31 @@ public void inProgressParallelInParallel() throws Exception {
         r.assertBuildStatusSuccess(r.waitForCompletion(b));
         sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
     }
+
+    @Test
+    public void inProgressParallelInParallelEvenNesting() throws Exception {
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "parallel(\n" +
+                "  'outerA': {\n" +
+                "    parallel('innerA': {\n" +
+                "       semaphore('innerA')\n" +
+                "    })\n" +
+                "  },\n" +
+                "  'outerB': {\n" +
+                "    parallel('innerB': {\n" +
+                "       semaphore('innerB')\n" +
+                "    })\n" +
+                "  }\n" +
+                ")", true));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("innerA/1", b);
+        SemaphoreStep.waitForStart("innerB/1", b);
+        ForkScanner scanner = new ForkScanner();
+        sanityTestIterationAndVisiter(b.getExecution().getCurrentHeads());
```

Comment:
This fails. The current behavior is that ForkScanner visits the two `semapshore` steps,  the `innerA` and `innerB` branches, and then the _outer_ `parallel` step, and the `FlowStartNode`.

Reply:
Oh and I have tested various combinations and this seems to be the simplest reproducer - 2 outer branches each with 1 inner branch. You can add more outer branches and the problem still occurs, but if you only have 1 outer branch things work fine. If you add more than 1 inner branch, the problem goes away for the outer branches that have more than 1 inner branch.

Reply:
Ah actually we can slightly simplify things, 2 outer branches with the first branch having 1 inner branch and the second branch not having nested parallelism also reproduces the issue.

Reply:
FTR https://github.com/jenkinsci/workflow-api-plugin/pull/302/checks?check_run_id=16190024918

```
ForkScanner should visit the same number of nodes as DepthFirstScanner expected:<10> but was:<6>
```

Reply:
[819d8a8](https://github.com/jenkinsci/workflow-api-plugin/pull/302/commits/819d8a89242aa123239f321c5566046727b69d98) should fix the issue.